strawberry/experimental/pydantic/nested/base.py

```diff
@@ -0,0 +1,241 @@
+            isclass(self.field.type_)
+            and issubclass(self.field.type_, pydantic.BaseModel)
+        ):
+            self.child_type: Tuple[Any, ...] = (self.field.type_,)
+        elif get_origin(self.field.type_) is Union and any(
+            _may_be_model(t) for t in get_args(self.field.type_)
+        ):
+            self.child_type = get_args(self.field.type_)
+
+    def _process_type(self, typ, idx: int = 0) -> object:
+        if isinstance(typ, ForwardRef):
+            return LazyForwardRefType[self.model, self.name, idx]  # type: ignore
+        elif isclass(typ) and issubclass(typ, pydantic.BaseModel):
+            return getattr(typ, "_strawberry_type", LazyModelType[typ])  # type: ignore
+        # Already a strawberry type
+        return typ
+
+    def _process_union(self, union_types: Tuple[Any, ...]) -> Union[Any, Any]:
+        """Build an Union type out of a type params.
+
+        Python < 3.11 don't allow the `typing.Union[*params]` syntax
+        so we build the union iteratively by pairs, using the following property :
+        Union[a, b, c] = Union[Union[a, b], c]
+        """
+        assert len(union_types) > 1
+
+        union: Union[Any, Any] = None
+        for i, (typ_a, typ_b) in enumerate(zip(union_types, union_types[1:])):
+            typ_a = self._process_type(typ_a, i)
+            typ_b = self._process_type(typ_b, i + 1)
+            union = (
+                Union[typ_a, typ_b]
+                if union is None
+                else Union[union, Union[typ_a, typ_b]]
+            )
+        return union
+
+    def get_strawberry_type(self) -> object:
+        none_in_union = False
+        if len(self.child_type) > 1:
+            strawberry_type = self._process_union(self.child_type)
+            none_in_union = any(t is None.__class__ for t in self.outer_types)
+        else:
+            strawberry_type = self._process_type(self.child_type[0])
+
+        # Rebuild outer types
+        for typ in self.outer_types[-1::-1]:
+            strawberry_type = typ[strawberry_type]
+
+        if none_in_union or not self.required:
+            strawberry_type = Optional[strawberry_type]
+
+        return strawberry_type
+
+    @property
+    def required(self) -> bool:
+        # Check outermost type as `self.model.field.required` can be of type Undefined
+        return (
+            getattr(get_type_hints(self.model)[self.name], "_name", None) != "Optional"
+            if isinstance(self.field.required, UndefinedType)
+            else self.field.required
+        )
+
+    @cached_property
+    def outer_types(self) -> List[Any]:
+        return extract_type_list(self.field.outer_type_)[0]
+
+
+def _may_be_model(typ: Any) -> bool:
+    """Whether or not the given type may leads to a Pydantic model."""
+    return (isclass(typ) and issubclass(typ, pydantic.BaseModel)) or isinstance(
+        typ, ForwardRef
+    )
+
+
+def extract_type_list(typ: Any) -> Tuple[List[Any], Tuple[Any, ...]]:
```

Comment:
Could all of this be making using of `StrawberryAnnotation`/`StrawberryType`? At a (cursory) glance, we're recreating some effort within the repo.

Reply:
Sure. I started this as a poc by overloading the pydantic decorator, without intending to open a PR so I suppose we integrate this deeper with strawberry.